File: jlpt/native_ios/Sources/Data/API/StudyAPI.swift
import Foundation

final class StudyAPI {
    private let client: APIClient
    init(client: APIClient) { self.client = client }

    func getVocabulary(level: String, page: Int, size: Int) async throws -> [VocabularyStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/vocabulary", query: pageQuery(level: level, page: page, size: size)))
    }

    func getKanji(level: String, page: Int, size: Int) async throws -> [KanjiStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/kanji", query: pageQuery(level: level, page: page, size: size)))
    }

    func getGrammar(level: String, page: Int, size: Int) async throws -> [GrammarStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/grammar", query: pageQuery(level: level, page: page, size: size)))
    }

    func getVocabularyDue(level: String, page: Int, size: Int) async throws -> [VocabularyStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/vocabulary/due", query: pageQuery(level: level, page: page, size: size)))
    }

    func getKanjiDue(level: String, page: Int, size: Int) async throws -> [KanjiStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/kanji/due", query: pageQuery(level: level, page: page, size: size)))
    }

    func getGrammarDue(level: String, page: Int, size: Int) async throws -> [GrammarStudyResponseDTO] {
        try await client.send(Endpoint(path: "api/study/grammar/due", query: pageQuery(level: level, page: page, size: size)))
    }

    func updateProgress(_ request: UpdateProgressRequestDTO) async throws {
        try await client.sendNoContent(Endpoint(
            path: "api/study/progress",
            method: .post,
            body: Endpoint.withJSONBody(request, encoder: client.encoder)
        ))
    }

    func review(_ request: ReviewRequestDTO) async throws -> ReviewResultResponseDTO {
        try await client.send(Endpoint(
            path: "api/study/review",
            method: .post,
            body: Endpoint.withJSONBody(request, encoder: client.encoder)
        ))
    }

    func getProgressSummary() async throws -> ProgressSummaryResponseDTO {
        try await client.send(Endpoint(path: "api/users/me/progress"))
    }

    private func pageQuery(level: String, page: Int, size: Int) -> [URLQueryItem] {
        [
            URLQueryItem(name: "level", value: level),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "size", value: String(size))
        ]
    }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/remote/api/StudyApi.kt
package com.jlpt.android.data.remote.api

import com.jlpt.android.data.remote.dto.GrammarStudyResponseDto
import com.jlpt.android.data.remote.dto.KanjiStudyResponseDto
import com.jlpt.android.data.remote.dto.ProgressSummaryResponseDto
import com.jlpt.android.data.remote.dto.ReviewRequestDto
import com.jlpt.android.data.remote.dto.ReviewResultResponseDto
import com.jlpt.android.data.remote.dto.UpdateProgressRequestDto
import com.jlpt.android.data.remote.dto.VocabularyStudyResponseDto
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

interface StudyApi {
    @GET("api/study/vocabulary")
    suspend fun getVocabulary(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<VocabularyStudyResponseDto>

    @GET("api/study/kanji")
    suspend fun getKanji(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<KanjiStudyResponseDto>

    @GET("api/study/grammar")
    suspend fun getGrammar(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<GrammarStudyResponseDto>

    @GET("api/study/vocabulary/due")
    suspend fun getVocabularyDue(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<VocabularyStudyResponseDto>

    @GET("api/study/kanji/due")
    suspend fun getKanjiDue(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<KanjiStudyResponseDto>

    @GET("api/study/grammar/due")
    suspend fun getGrammarDue(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): List<GrammarStudyResponseDto>

    @POST("api/study/progress")
    suspend fun updateProgress(@Body request: UpdateProgressRequestDto)

    @POST("api/study/review")
    suspend fun review(@Body request: ReviewRequestDto): ReviewResultResponseDto

    @GET("api/users/me/progress")
    suspend fun getProgressSummary(): ProgressSummaryResponseDto
}
